### ISSUE TITLE
Player often takes lock UnloadedUpdate on vessels in different subspace

### ISSUE THREAD

**Issue by progfz** [LunaMultiplayer/LunaMultiplayer#292]
Intro:

Me, my friend and his girlfriend tried to start playing KSP with LMP. All went good at start but then we started to encounter weird desync behaviours in various forms, to state a few there was disappearing and teleporting ships and crew and ghost ships existing only on one client after revert or recovery.

As I'm more or less knowledgeable about programming, I started investigating this issue. First I eliminated other mods influence, leaving vanilla client with LMP only, issue persisted. Then after some experiments and a bit of source code digging formed what I'm logging here.

I'll add more info as I further investigate.
Some of the issues we encountered could be not related to this or I could be completely wrong in my assumptions. For example, I don't know yet how to explain vessel instantly teleporting from mun orbit to kerbin orbit or vessels disappering when approached during landing as this situations are harder to reproduce.

------------------------------------------------------------------------------
------------------------------------------------------------------------------
------------------------------------------------------------------------------

OS: 
Windows 7
------------------------------------------------------------------------------

KSPVersion: 
1.7
------------------------------------------------------------------------------

LMP version (include the build number if you are using a nighly build): 
Build 0.21.0.210 from github
------------------------------------------------------------------------------

MODS (add link to download them): 
No mods, just vanilla and LMP
------------------------------------------------------------------------------

Expected behaviour: 

OptionA: Player should not try to get UnloadedUpdate lock on vessels in different subspace. Taking this lock on vessels from different subspace could lead to vessels derailing and even destruction. Not taking this lock, though, could lead to outdated orbital info about vessel.

OptionB: Player should have "priority lock" on vessels player launched or controlled recently and other players should not take UnloadedUpdate on vessels with "priority lock" from other players. Taking direct control of vessel should override priority lock to a new value, taking normal Update lock could override "priority lock" too. This approach could lead to the same situation with outdated orbital info about vessel as OptionA but more cleanly states that player with "priority lock" is responsible for taking UnloadedUpdate lock whenever applicable instead of random players sitting in control station.

OptionC: according to playtesting with three players in locally built copy of LMP, UnloadedUpdate could be disabled completely with acceptable degree of side effects.

------------------------------------------------------------------------------

Current behaviour: 

Player could automatically take UnloadedUpdate lock on vessels in different subspace, potentially leading to destruction or derailing of vessel due to difference in planet positions.
------------------------------------------------------------------------------

Steps to reproduce: 

1. Have two players on server. Two windows on single machine works fine as well though may require reducing graphics settings depending on hardware.
2. PlayerA and PlayerB should be in different subspaces. PlayerA should be ahead of PlayerB in time and just sitting in control center or VAB.
3. PlayerB launches vessel. Loocks aquired, vessel saved.
4. PlayerB returns to control center leaving vessel in world. At this moment all locks on this vessel are released.
5. PlayerA goes to tracking station. At this moment PlayerA acquires lock on vessel that PlayerB launched despite it being in different subspace.
6. Depending on time difference and positioning nothing will happen at all or vessel will bump into surface and disappear or teleport somewhere due to derailing.

Bonus: at step 3 vessel could be left near mun or minmus, potentially causing it to derail or disappear if time difference and positioning are "good" enough to call space-time kraken.

**Comment by progfz** [comment]
Update:
Did local build. Completely disabled UnloadedUpdate locks. Results of playtesting did show that most of the disappearance and derailing issues are gone with just minor side effects.
Will further investigate anyway in hopes to create solution that's more in line with current mod architecture but it will take time.

**Comment by 40339110** [comment]
> Update:
> Did local build. Completely disabled UnloadedUpdate locks. Results of playtesting did show that most of the disappearance and derailing issues are gone with just minor side effects.
> Will further investigate anyway in hopes to create solution that's more in line with current mod architecture but it will take time.

@progfz Could you give a general overview of exactly how you did this? Having the same issues with ships reappearing in odd places and disappearing altogether, and I cant find any other reports of the issue.

**Comment by gavazquez** [comment]
This bug is pulling my hair off. Everybody seems to expierence it at some point but I can't manage to reproduce it.

I did all the steps and it doesn't happen :/

1: Connect player A and player B
2: Player A warps 15 days forward while staying at KSC
3: Player B launch and cheat to set orbit on mun at 68K
4: Player B leaves to KSC
5: Player A goes to Track station

After all those steps everything is still normal....

**Comment by progfz** [comment]
@gavazquez
My reproduction steps are for players taking authority over unloaded ships they should not take, observed in server output. Full reproduction steps with visible changes to ships are a bit more complicated as they involve calculating exact time to warp depending on orbit of involved vessel.
Checklist helpful for full reproduction:
- PlayerA should be in different timespace, can't be reproduced in same timespace.
- PlayerA should take unloaded updates lock over vessel launched by PlayerB from different timespace, can't be reproduced if unloaded lock is not aquired.
- PlayerA may need to timewarp some more after getting unloaded update lock on vessel.
- Viewpoint of both players should be investigated and syncing time after final timewarp may be needed to apply changes to orbit too. There are no such steps in my simple reproduction as locks are visible in server log and this issue I logged is based around locks, not around visible side-effects.
- More unstable orbits help reproducing too, vessel near edge of SOI is more easily pulled by this bug than one solidly sitting in SOI as smaller shifts to orbit are needed in this case to get noticeable side effects.

In addition to simple derailing there are two more most prominent side-effects, but I did not investigate them that much
- Death warp: vessel is heading on crash trajectory to planet, left uncontrolled for a while, player takes unloaded lock over vessel and timewarps to future where vessel is destroyed, propagating destruction to all players even if player that launched vessel is still in the same timespace with that vessel just not in control. May be unrelated to unloaded updates, did not have time to investigate if this still happens with unloaded updates disabled.
- Vessel replacement: vessel sometimes is spontaneously replaced with another vessel or asteroid. I did not thoroughly investigate this one either but one thing I'm sure about - it disappears completely with unloaded updates disabled as we did not encounter it anymore.

I could recommend temporary solution as quick-fix - to disable unloaded updates completely for now, via adding settings variable or hardcoded in quick-fix release, this will greatly improve situation for players. Any side-effects of disabling unloaded updates are less harmful than current situation based on my playtesting with friends when I had time for it.
